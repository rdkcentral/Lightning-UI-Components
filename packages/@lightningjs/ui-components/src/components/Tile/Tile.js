/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import Artwork from '../Artwork';
import Badge from '../Badge';
import Checkbox from '../Checkbox';
import Label from '../Label';
import MetadataTile from '../MetadataTile';
import ProgressBar from '../ProgressBar';
import * as styles from './Tile.styles';
import Surface from '../Surface';

export default class Tile extends Surface {
  static get __componentName() {
    return 'Tile';
  }

  static get __themeStyle() {
    return styles;
  }

  static _template() {
    return {
      ...super._template(), // Make sure surface elements are patched
      Tile: {
        Artwork: {
          type: Artwork,
          signals: {
            imageLoaded: '_imageLoaded'
          },
          mount: 0.5
        },
        Content: {
          mount: 0.5
        }
      }
    };
  }

  static get properties() {
    return [
      'artwork',
      'circle',
      'badge',
      'checkbox',
      'metadata',
      'metadataLocation',
      'persistentMetadata',
      'progressBar',
      'label',
      'src'
    ];
  }

  static get tags() {
    return [
      ...super.tags,
      'Artwork',
      'Content',
      'Tile',
      { name: 'Badge', path: 'Content.Badge' },
      { name: 'Checkbox', path: 'Content.Checkbox' },
      { name: 'Metadata', path: 'Content.Metadata' },
      { name: 'ProgressBar', path: 'Content.ProgressBar' },
      { name: 'Label', path: 'Content.Label' }
    ];
  }

  /**
   * When metadata is displayed below the Tile we need a way to tell the containing columns that it exists
   * in order for it to layout properly. This approach will not however show up in the DOM inspector
   */

  set announce(announce) {
    super.announce = announce;
  }

  get announce() {
    return (
      this._announce || [
        this._Metadata && this._Metadata.announce,
        this._Badge && this._Badge.announce,
        this._Label && this._Label.announce,
        this._ProgressBar && this._ProgressBar.announce
      ]
    );
  }

  _update() {
    super._update();
    this._updateTileColor();
    this._updateContent();
    this._updateArtwork();
    this._updateBadge();
    this._updateLabel();
    this._updateCheckbox();
    this._updateProgressBar();
    this._updateMetadata();
  }

  /* ------------------------------ General Tile Layout ------------------------------ */

  set h(v) {
    super.h = v;
  }

  get h() {
    return this.metadataLocation !== 'inset'
      ? super.h + ((this._Metadata && this._Metadata.h) || 0)
      : super.h;
  }

  get innerH() {
    return this._h; // Ensure that surface respects the correct height when metadata is displayed below
  }

  get _gradient() {
    if (this._isCircleLayout) return false;
    return Boolean(
      this._isInsetMetadata &&
        this._hasMetadata &&
        (this._shouldShowMetadata || this._shouldShowBadgeLabel)
    );
  }

  get _isCircleLayout() {
    return Boolean(this._itemLayout && this._itemLayout.circle);
  }

  get _foregroundDefaultWidth() {
    return parseFloat(this._w / this._h).toFixed(2) ===
      parseFloat(16 / 9).toFixed(2)
      ? this.innerW * 0.5
      : this.innerW * 0.75;
  }

  _updateTileColor() {
    this._Tile.alpha = this.style.alpha;
  }

  /* ------------------------------ Artwork ------------------------------ */
  _updateArtwork() {
    // ensure a nested artwork src takes precedence over the class's src setter,
    // but that if src is undefined in both the setter and artwork object,
    // we don't incorrectly pass "src: undefined" to the Artwork component)
    this._Artwork.patch({
      h: this._h,
      w: this._w,
      x: this._w / 2,
      y: this._h / 2,
      src: this.src,
      ...(this.artwork || {}),
      style: {
        radius: this.style?.radius,
        ...this.artwork?.style
      },
      gradient: this._gradient,
      shouldScale: this._isFocusedMode
    });
  }

  _getSrc() {
    return (this.artwork && this.artwork.src) || this._src;
  }

  _imageLoaded() {
    this._Background.alpha = 0; // Since the image is loaded the surface does not need to be shown
  }

  /* ------------------------------ Badge ------------------------------ */
  _updateBadge() {
    const badgePatch = {
      ...this.badge,
      x: this.style.paddingX,
      y: this.style.paddingY,
      alpha:
        !this._persistentMetadata || !this.badge?.title || this._isCircleLayout
          ? 0.001
          : 1
    };
    if (!this._Badge) {
      this._Content.patch({
        Badge: {
          type: Badge,
          ...badgePatch,
          signals: {
            loadedBadge: '_updateBadge'
          }
        }
      });
      return;
    }

    this.applySmooth(this._Badge, badgePatch, {
      ...badgePatch,
      ...this._badgeLabelTransitions // Badge and Label should animate in with the same values
    });
  }

  /* ------------------------------ Label  ------------------------------ */
  _updateLabel() {
    if (!this.label?.title || this._isCircleLayout) {
      if (this._Label) {
        this._Content.patch({
          Label: undefined
        });
      }
      return;
    }
    const labelPatch = {
      ...this.label,
      x: this._w - this.style.paddingX,
      y: this.style.paddingY,
      alpha:
        !this._persistentMetadata || this._isCircleLayout || !this.label.title
          ? 0.001
          : 1
    };

    if (!this._Label) {
      this._Content.patch({
        Label: {
          type: Label,
          mountX: 1,
          ...labelPatch,
          signals: {
            loadedLabel: '_updateLabel'
          }
        }
      });
      return;
    }

    this.applySmooth(this._Label, labelPatch, {
      ...labelPatch,
      x: [
        labelPatch.x,
        this._shouldShowBadgeLabel
          ? this.style.animationEntrance
          : this.style.animationExit
      ],
      ...this._badgeLabelTransitions
    });
  }
  /* ------------------------------ Badge & Label ------------------------------ */

  // Badge and Label should animate in with the same values
  get _shouldShowBadgeLabel() {
    return this._persistentMetadata || this._isFocusedMode;
  }
  get _badgeLabelTransitions() {
    return {
      y: [
        this._shouldShowBadgeLabel ? this.style.paddingY : 0,
        this.__shouldShowBadgeLabel
          ? this.style.animationEntrance
          : this.style.animationExit
      ],
      alpha: [
        this._shouldShowBadgeLabel && !this._isCircleLayout ? 1 : 0.001,
        this._shouldShowBadgeLabel
          ? this.style.animationEntrance
          : this.style.animationExit
      ]
    };
  }

  /* ------------------------------ Checkbox ------------------------------ */
  _updateCheckbox() {
    // Remove Checkbox if no longer required
    if (
      !(typeof this.checkbox?.checked === 'boolean' && this.checkbox.checked) ||
      this._isCircleLayout
    ) {
      if (this._Checkbox) {
        this._Content.patch({
          Checkbox: undefined
        });
      }
      return;
    }

    const checkboxPatch = {
      ...this.checkbox,
      x: this._w - this.style.paddingX,
      y: this._h - this.style.paddingY
    };

    if (!this._Checkbox) {
      this._Content.patch({
        Checkbox: {
          ...checkboxPatch,
          type: Checkbox,
          mount: 1
        }
      });
      return;
    }

    this.applySmooth(this._Checkbox, checkboxPatch);
  }
  /* ------------------------------ Progress Bar ------------------------------ */

  get _progressBarHeight() {
    return (
      (this._ProgressBar &&
        this._ProgressBar._getTransition('h')._targetValue +
          this.style.paddingY) ||
      0
    );
  }
  _updateProgressBar() {
    // Remove ProgressBar if no longer required
    if (
      !(
        typeof this.progressBar?.progress === 'number' &&
        this.progressBar.progress
      ) ||
      this._isCircleLayout
    ) {
      if (this._ProgressBar) {
        if (this.shouldSmooth) {
          this._ProgressBar._getTransition('alpha').once('finish', () => {
            this._removeProgressBar();
          });
          this._ProgressBar.smooth = { alpha: 0 };
        } else {
          this._removeProgressBar();
        }
      }
      return;
    }
    // --------------------------------------//
    if (this.progressBar.progress > 0) {
      const progressPatch = {
        ...this.progressBar,
        w: this._w - this.style.paddingX * 2,
        x: this._w / 2,
        y: this._h - this.style.paddingYProgress
      };

      if (!this._ProgressBar) {
        this._Content.patch({
          ProgressBar: {
            ...progressPatch,
            type: ProgressBar,
            mountX: 0.5,
            mountY: 1,
            alpha: this._hasMetadata && this.shouldSmooth ? 0.001 : 1
          }
        });

        if (this.shouldSmooth) {
          this._ProgressBar.smooth = {
            alpha: [
              1,
              { delay: this.style.animationEntrance.duration } // Wait for metadata to animate in
            ]
          };
        }
        return;
      }

      // TODO: See if we need to add animation to every property individually or can set parent
      this.applySmooth(
        this._ProgressBar,
        progressPatch,
        Object.keys(progressPatch).reduce((acc, prop) => {
          acc[prop] = [
            progressPatch[prop],
            this._isFocusedMode
              ? this.style.animationEntrance
              : this.style.animationExit
          ];
          return acc;
        }, {})
      );
    }
  }
  _removeProgressBar() {
    this._Content.patch({ ProgressBar: undefined });
    this._updateMetadata();
  }
  /* ------------------------------ Metadata  ------------------------------ */

  get _shouldShowMetadata() {
    return this._persistentMetadata || this._isFocusedMode;
  }

  get _isInsetMetadata() {
    return this._metadataLocation === 'inset';
  }
  // called in animateMetadata
  get _metadataTransitions() {
    return {
      y: [
        this._shouldShowMetadata || this._isInsetMetadata
          ? this._metadataY
          : this._h + this.style.paddingY,
        this._shouldShowMetadata
          ? this.style.animationEntrance
          : this.style.animationExit
      ],
      alpha: [
        this._shouldShowMetadata ? 1 : 0.001,
        this._shouldShowMetadata
          ? this.style.animationEntrance
          : this.style.animationExit
      ]
    };
  }

  get _hasMetadata() {
    return MetadataTile.properties.some(
      prop => this.metadata && this.metadata[prop]
    );
  }

  get _metadataY() {
    return this._isInsetMetadata
      ? this._h - this.style.paddingY - this._progressBarHeight
      : this._h + this.style.paddingY;
  }

  get _metadataPatch() {
    return {
      alpha: this._hasMetadata && this._shouldShowMetadata ? 1 : 0.001,
      mountX: 0.5,
      mountY: this._isInsetMetadata ? 1 : 0,
      marquee: this._isFocusedMode,
      w: this._w - this.style.paddingX * 2,
      x: this._w / 2,
      y: !(this._isInsetMetadata && this._isFocusedMode)
        ? this._metadataY
        : this._h + this.style.paddingY,
      ...(this.metadata || {})
    };
  }

  _updateMetadata() {
    if (
      !this._hasMetadata ||
      (this._isCircleLayout && this._metadataLocation === 'inset')
    ) {
      return;
    }

    if (!this._persistentMetadata && !this._isFocusedMode) {
      this._animateMetadata();
      return;
    }

    if (!this._Metadata) {
      // Patch in Metadata for the first time
      this._Content.patch({
        Metadata: {
          type: MetadataTile,
          signals: {
            updateComponentDimensions: '_metadataLoaded'
          },
          ...this._metadataPatch
        }
      });

      return;
    }

    this._Metadata.patch(this._metadataPatch); // Metadata should never be patched with smooth

    this._animateMetadata();
  }

  _animateMetadata() {
    if (!this._Metadata) return;
    this.applySmooth(
      this._Metadata,
      this._metadataPatch,
      this._metadataTransitions
    );

    if (!this._isFocusedMode) {
      this._resetMarqueeAnimation();
    }
  }

  // signal in updateMetadata
  _metadataLoaded() {
    this._animateMetadata();
    if (this.metadataLocation !== 'inset') this.fireAncestors('$itemChanged'); // Send event to columns/rows that the height has been updated since metadata will be displayed below the Tile
  }

  /* ------------------------------ ?????  ------------------------------ */

  _updateContent() {
    const itemContainerPatch = {
      h: this._h,
      w: this._w,
      x: this._w / 2,
      y: this._h / 2
    };

    // Make sure container animates with same values as badge, label, and metadata
    this.applySmooth(
      this._Content,
      itemContainerPatch,
      Object.keys(itemContainerPatch).reduce((acc, prop) => {
        acc[prop] = [
          itemContainerPatch[prop],
          this._isFocusedMode // this is attaching transitions to props in itemContainerPatch
            ? this.style.animationEntrance
            : this.style.animationExit
        ];
        return acc;
      }, {})
    );
  }

  _resetMarqueeAnimation() {
    const alphaTransition = this._Metadata._getTransition('alpha');
    if (alphaTransition) {
      alphaTransition.on('finish', () => {
        if (this._Metadata) {
          this._Metadata.resetMarquee();
        }
      });
    } else {
      this._Metadata.resetMarquee();
    }
  }
}
