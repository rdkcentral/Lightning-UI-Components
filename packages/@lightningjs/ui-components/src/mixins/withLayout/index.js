/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { context } from '../../globals';
import { stringifyCompare, getDimensions } from '../../utils';

export default function withLayout(Base) {
  return class extends Base {
    _construct() {
      super._construct && super._construct();
      this._previousDimensionData = null; // Store the previous dimension data to compare next time _updateItemLayout is called
    }

    get itemLayout() {
      return this._itemLayout;
    }

    set itemLayout(v) {
      const componentName =
        this.constructor._componentName || this.constructor.name;
      let itemLayout;
      if (v) {
        itemLayout = JSON.parse(
          JSON.stringify(v, (k, v) => {
            if (k !== 'circle' && v < 0) {
              context.error(
                `itemLayout for ${componentName} received an invalid value of ${v} for ${k}`
              );
              return;
            } else if (k === 'circle') {
              return Boolean(v);
            }
            return v;
          })
        );
      }

      if (!stringifyCompare(this._itemLayout, itemLayout)) {
        if (itemLayout && !itemLayout.upCount) {
          this._originalW = this.w;
          this._originalH = this.h;
          this._itemLayout = {
            w: this._originalW,
            h: this._originalH,
            ...itemLayout
          };
        } else {
          this._itemLayout = itemLayout;
        }

        this._updateItemLayout();
      }
    }

    _allowUpdate() {
      const {
        w = '',
        h = '',
        circle = '',
        ratioX = '',
        ratioY = '',
        upCount = ''
      } = this._itemLayout || {};
      if (!this.style) {
        return false;
      }
      // Store a string representation of all values from previous update to compare and decide if layout should change
      const layoutString =
        Object.values(context.theme.layout).join('') +
        `${w}${h}${circle ? 1 : 0}${ratioX}${ratioY}${upCount}`;
      if (layoutString !== this._previousDimensionData) {
        this._previousDimensionData = layoutString;
        return true;
      }
      return false;
    }

    _updateItemLayout() {
      if (!this._allowUpdate()) return;
      const { w, h } = getDimensions(this.theme, this._itemLayout);
      if (h || w) {
        // If there is not enough information passed in args to calculate item size
        const width = context.theme.layout.screenW;
        const height = context.theme.layout.screenH;
        const calculatedWidth = w || h * (width / height);
        const calculatedHeight = h || w * (height / width);
        this.w =
          this._itemLayout && this._itemLayout.circle
            ? calculatedHeight
            : calculatedWidth; // Width must be set first in order for Cards to be displayed properly
        this.h = calculatedHeight;
        if (this._itemLayout && this._itemLayout.circle && this.style) {
          if (-1 < this._componentStyleSourceProps.indexOf('radius')) {
            this._circleSet = true;
            this._originalRadius = this.style.radius;
            this.style = {
              ...this.style,
              radius: calculatedHeight / 2
            };
          }
        } else if (this._circleSet) {
          // Restore to it's original state
          this.style = {
            ...this.style,
            radius: this._originalRadius
          };
          this._originalRadius = undefined;
          this._circleSet = false;
        }
        this.queueRequestUpdate && this.queueRequestUpdate();
        this.fireAncestors('$itemChanged');
      }
    }
  };
}
