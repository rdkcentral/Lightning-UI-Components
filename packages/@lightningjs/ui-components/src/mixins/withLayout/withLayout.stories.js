/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import lng from '@lightningjs/core';
import mdx from './withLayout.mdx';
import { Tile } from '../../components';
import { CATEGORIES } from '../../docs/constants';

export default {
  title: `${CATEGORIES[512]}/withLayout`,
  parameters: {
    docs: {
      page: mdx
    }
  }
};

export const withLayout = () => {
  return class withLayout extends lng.Component {
    static _template() {
      return {
        Item: {
          type: Tile,
          artwork: {
            src: 'https://image.tmdb.org/t/p/w500/frwl2zBNAl5ZbFDJGoJv0mYo0rF.jpg',
            mode: 'contain'
          }
        }
      };
    }
  };
};

withLayout.storyName = 'withLayout';

withLayout.args = {
  ratioX: 16,
  ratioY: 9,
  upCount: 3,
  circle: false
};

withLayout.argTypes = {
  ratioX: {
    control: { type: 'number', min: 0 },
    description: 'The units of x resolution relative to the screen height',
    table: { defaultValue: { summary: 0 } }
  },
  ratioY: {
    control: { type: 'number', min: 0 },
    description: 'The units of y resolution relative to the screen width',
    table: { defaultValue: { summary: 0 } }
  },
  upCount: {
    control: { type: 'number', min: 0 },
    description:
      'The number of items that should be displayed within screen bounds',
    table: { defaultValue: { summary: 0 } }
  },
  circle: {
    control: 'boolean',
    description: 'Boolean changing format to circle if true',
    table: { defaultValue: { summary: false } }
  }
};

withLayout.parameters = {
  argActions: Object.keys(withLayout.args).reduce((acc, curr) => {
    return {
      ...acc,
      [curr]: (val, component) => {
        component.tag('Item').itemLayout = {
          ...component.tag('Item').itemLayout,
          [curr]: val
        };
      }
    };
  }, {})
};
