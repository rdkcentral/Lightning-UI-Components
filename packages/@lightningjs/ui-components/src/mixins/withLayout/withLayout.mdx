import { Meta } from '@storybook/blocks';
import * as withLayoutStories from './withLayout.stories';

<Meta of={withLayoutStories} />

# withLayout

Use `withLayout` to autotomatically size an item according to screen dimensions, upCount and/or set explicitly.

## Source

https://github.com/rdkcentral/Lightning-UI-Components/blob/develop/packages/@lightningjs/ui-components/src/mixins/withLayout/index.js

## Usage

Let's use an example to see how we can use `withLayout`. Let's say I have an a `Tile` component on screen that should be sized according the screen size:

```js
import lng from '@lightningjs/core';
import { withLayout } from '@lightningjs/ui-components';

class Basic extends lng.Component {
  static _template() {
    return {
      Tile: {
        type: withLayout(Tile),
        src: parks,
        itemLayout: {
          ratioX: 16,
          ratioY: 9,
          upCount: 5
        }
      }
    };
  }
}
```

If I want to explicitly set a height but the only other information is the screen dimensions, I can do that by adding these parameters to the itemLayout object.

```js
import lng from '@lightningjs/core';
import { withLayout } from '@lightningjs/ui-components';

class Basic extends lng.Component {
  static _template() {
    return {
      Tile: {
        type: withLayout(Tile),
        src: parks,
        itemLayout: {
          ratioX: 16,
          ratioY: 9,
          h: 180
        }
      }
    };
  }
}
```

If I am unsure of the width, I can autosize by adding a height and upCount properties to the itemLayout object.

```js
import lng from '@lightningjs/core';
import { withLayout } from '@lightningjs/ui-components';

class Basic extends lng.Component {
  static _template() {
    return {
      Tile: {
        type: withLayout(Tile),
        src: parks,
        itemLayout: {
          h: 320,
          upCount: 4
        }
      }
    };
  }
}
```

**NOTE**: A combination of properties is required:

`h and w` (If explicitly set, it may be uncessary to use the withLayout pattern as these properties can be set on the component),
`h and ratioX and ratioY`,
`ratioX and ratioY and upCount`,
`h and upCount`,
`h`, and finally `w`

Auto sizing of the item will happen based on the above priority. So if the `w and h` are passed no other calculations will occur.
If the mixin is unable to calculate the height OR width from the passed parameters. The height will default 9/16 of the width. Inversely the width will default to 16/9 of the height.
If the mixin is unable to calculate height AND width from the passed parameters. No autosize will occur.

### Properties

| name       | type   | required | default | description                                                                  |
| ---------- | ------ | -------- | ------- | ---------------------------------------------------------------------------- |
| itemLayout | object | true     | -       | the auto sizing properties that will override explicitly set h/w for an item |

### itemLayout

| name    | type   | required | default | description                                                                                                                                                              |
| ------- | ------ | -------- | ------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| w       | number | false    | 0       | explicitly set the width of an item                                                                                                                                      |
| h       | number | false    | 0       | explicitly set the height of an item                                                                                                                                     |
| ratioX  | number | false    | 0       | units of x resolution relative to the screen height, should be a posititve integer                                                                                       |
| ratioY  | number | false    | 0       | units of y resolution relative to the screen width, should be a posititve integer                                                                                        |
| upCount | number | false    | 0       | the number of items that should be displayed within screen bounds. Additional items will run off-screen or outside of specified row width, should be a posititve integer |
