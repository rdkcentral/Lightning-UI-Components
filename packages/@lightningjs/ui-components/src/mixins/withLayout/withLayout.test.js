/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { makeCreateComponent } from '@lightningjs/ui-components-test-utils';
import withLayout from '.';
import Tile from '../../components/Tile';
import context from '../../globals/context';
import { jest } from '@jest/globals';
const createTileWithLayout = props =>
  makeCreateComponent(withLayout(Tile))(props);

describe('withLayout', () => {
  let tileWithLayout, testRenderer;
  const w = 320;
  const h = 180;
  const ratioY = 9;
  const ratioX = 16;
  const upCount = 5;

  beforeEach(() => {
    [tileWithLayout, testRenderer] = createTileWithLayout();
  });

  afterEach(() => {
    tileWithLayout = null;
    testRenderer = null;
  });

  it('renders', () => {
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it('should explicity set height, width', () => {
    tileWithLayout.itemLayout = { h, w };
    testRenderer.update();
    expect(tileWithLayout.h).toBe(h);
    expect(tileWithLayout.w).toBe(w);
    expect(tileWithLayout.itemLayout).toStrictEqual({ h: 180, w: 320 });
  });

  it('should set width from height and screen dimension', () => {
    tileWithLayout.itemLayout = { h, ratioX, ratioY };
    testRenderer.update();
    expect(tileWithLayout.h).toBe(h);
    expect(tileWithLayout.w).toBe((h * 16) / 9);
  });

  it('should set width from height and upCount', () => {
    tileWithLayout.itemLayout = { h, upCount };
    testRenderer.update();
    expect(tileWithLayout.w).toBeDefined();
    expect(tileWithLayout.w).not.toBe(0);
  });

  it('should set height and width from screen dimensions and upCount', () => {
    tileWithLayout.itemLayout = { ratioX, ratioY, upCount };
    testRenderer.update();
    expect(tileWithLayout.w).toBeDefined();
    expect(tileWithLayout.w).not.toBe(0);
  });

  it('should set height from width and default ratio', () => {
    tileWithLayout.itemLayout = { w };
    testRenderer.update();
    expect(tileWithLayout.h).toBeDefined();
    expect(tileWithLayout.h).toBe(
      (w * context.theme.layout.screenH) / context.theme.layout.screenW
    );
  });

  it('should set width from height and default ratio', () => {
    tileWithLayout.itemLayout = { h };
    testRenderer.update();
    expect(tileWithLayout.w).toBeDefined();
    expect(tileWithLayout.w).toBe(
      (h * context.theme.layout.screenW) / context.theme.layout.screenH
    );
  });

  it('should throw an error if a negative value is used', () => {
    const errorLog = jest.spyOn(global.console, 'error');
    tileWithLayout.itemLayout = { ratioX, ratioY, h, upCount };
    testRenderer.update();
    tileWithLayout.itemLayout.ratioX = -20;

    expect(errorLog).toHaveBeenCalled();
    // expect(logSpy).toHaveBeenCalledWith(
    //   'itemLayout for Tile recieved an invaild value of -16 for ratioX'
    // );
    expect(tileWithLayout.itemLayout.ratioX).toBe(16);
    //logSpy.console.error.mockRestore(); // restores original console log function
  });
});
