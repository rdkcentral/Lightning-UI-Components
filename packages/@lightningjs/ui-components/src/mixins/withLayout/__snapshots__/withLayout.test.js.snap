// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withLayout renders 1`] = `
{
  "Component": {
    "Background": {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Background",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "Background",
      ],
      "texture": {
        "type": "StaticCanvasTexture",
      },
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
    "Tile": {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": {
        "Artwork": {
          "active": false,
          "alpha": 0.001,
          "attached": true,
          "boundsMargin": null,
          "children": {
            "Image": {
              "active": false,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 0,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Image",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": [
                "Image",
              ],
              "texture": {
                "type": "CustomImageTexture",
              },
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 1,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0.5,
          "mountX": 0.5,
          "mountY": 0.5,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Artwork",
          "renderOfScreen": undefined,
          "renderToTexture": true,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": [
            "Artwork",
          ],
          "type": "Artwork",
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "Content": {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "isComponent": undefined,
          "mount": 0.5,
          "mountX": 0.5,
          "mountY": 0.5,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Content",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": [
            "Content",
          ],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "ProviderLogo": {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294506490,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "ProviderLogo",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": [
            "ProviderLogo",
          ],
          "type": "Icon",
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Tile",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "Tile",
      ],
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;
